Problem serializing zipkin2.Span type with Kryo

Kryo doesn't support Google's AutoValue extensions that zipkin2.Span uses. 

So the following error is seen during serialization:
```
Caused by: com.esotericsoftware.kryo.KryoException: java.lang.UnsupportedOperationException
Serialization trace:
annotations (zipkin2.AutoValue_Span)
        at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:528)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:761)
        at com.twitter.chill.TraversableSerializer.read(Traversable.scala:43)
        at com.twitter.chill.TraversableSerializer.read(Traversable.scala:21)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:679)
        at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:106)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:528)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:657)
....
Caused by: java.lang.UnsupportedOperationException
        at java.util.Collections$UnmodifiableCollection.add(Collections.java:1055)
        at com.esotericsoftware.kryo.serializers.CollectionSerializer.read(CollectionSerializer.java:116)
        at com.esotericsoftware.kryo.serializers.CollectionSerializer.read(CollectionSerializer.java:22)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:679)
        at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:106)
```

Kryo version: 2.24.0

--- a/benchmarks/pom.xml
+++ b/benchmarks/pom.xml
@@ -48,6 +48,11 @@
     </dependency>
 
     <dependency>
+      <groupId>com.esotericsoftware.kryo</groupId>
+      <artifactId>kryo</artifactId>
+    </dependency>
+
+    <dependency>
       <groupId>org.apache.thrift</groupId>
       <artifactId>libthrift</artifactId>
       <version>${libthrift.version}</version>

--- a/benchmarks/src/main/java/zipkin/benchmarks/CodecBenchmarks.java
+++ b/benchmarks/src/main/java/zipkin/benchmarks/CodecBenchmarks.java
@@ -1,5 +1,5 @@
 /**
- * Copyright 2015-2017 The OpenZipkin Authors
+ * Copyright 2015-2018 The OpenZipkin Authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
  * in compliance with the License. You may obtain a copy of the License at
@@ -13,6 +13,10 @@
  */
 package zipkin.benchmarks;
 
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import com.esotericsoftware.kryo.serializers.JavaSerializer;
 import com.google.common.io.ByteStreams;
 import com.twitter.zipkin.thriftjava.Annotation;
 import com.twitter.zipkin.thriftjava.BinaryAnnotation;
@@ -160,6 +164,29 @@ public class CodecBenchmarks {
   static final Span zipkin2 = SpanBytesDecoder.JSON_V2.decodeOne(zipkin2Json);
   static final List<Span> tenSpan2s = Collections.nCopies(10, zipkin2);
   static final byte[] tenSpan2sJson = SpanBytesEncoder.JSON_V2.encodeList(tenSpan2s);
+  static final Kryo kryo = new Kryo();
+  static final byte[] zipkin2Serialized;
+  static {
+    kryo.register(Span.class, new JavaSerializer());
+    Output output = new Output(4096);
+    kryo.writeObject(output, zipkin2);
+    output.flush();
+    zipkin2Serialized = output.getBuffer();
+  }
+
+  /** manually implemented with json so not as slow as normal java */
+  @Benchmark
+  public Span readClientSpan_java_zipkin2() {
+    return kryo.readObject(new Input(zipkin2Serialized), Span.class);
+  }
+
+  @Benchmark
+  public byte[] writeClientSpan_java_zipkin2() {
+    Output output = new Output(zipkin2Serialized.length);
+    kryo.writeObject(output, zipkin2);
+    output.flush();
+    return output.getBuffer();
+  }
 
   @Benchmark
   public Span readClientSpan_json_zipkin2() {
@@ -264,7 +291,7 @@ public class CodecBenchmarks {
   // Convenience main entry-point
   public static void main(String[] args) throws RunnerException {
     Options opt = new OptionsBuilder()
-        .include(".*" + CodecBenchmarks.class.getSimpleName() + ".*readClientSpan_json_zipkin2")
+        .include(".*" + CodecBenchmarks.class.getSimpleName() + ".*kryo_zipkin2")
         .build();
 
     new Runner(opt).run();

--- a/pom.xml
+++ b/pom.xml
@@ -365,6 +365,13 @@
         <version>1.5.3</version>
       </dependency>
 
+      <!-- for testing flink -->
+      <dependency>
+        <groupId>com.esotericsoftware.kryo</groupId>
+        <artifactId>kryo</artifactId>
+        <version>2.24.0</version>
+      </dependency>
+
       <dependency>
         <groupId>junit</groupId>
         <artifactId>junit</artifactId>

--- a/zipkin2/pom.xml
+++ b/zipkin2/pom.xml
@@ -79,6 +79,11 @@
       <version>${powermock.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.esotericsoftware.kryo</groupId>
+      <artifactId>kryo</artifactId>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
   <build>

--- a/zipkin2/src/main/java/zipkin2/Annotation.java
+++ b/zipkin2/src/main/java/zipkin2/Annotation.java
@@ -1,5 +1,5 @@
 /**
- * Copyright 2015-2017 The OpenZipkin Authors
+ * Copyright 2015-2018 The OpenZipkin Authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
  * in compliance with the License. You may obtain a copy of the License at
@@ -13,21 +13,22 @@
  */
 package zipkin2;
 
-import com.google.auto.value.AutoValue;
+import java.io.ObjectStreamException;
 import java.io.Serializable;
+import java.io.StreamCorruptedException;
 
 /**
  * Associates an event that explains latency with a timestamp.
  *
  * <p>Unlike log statements, annotations are often codes: Ex. {@code cache.miss}.
  */
-@AutoValue
 //@Immutable
-public abstract class Annotation implements Comparable<Annotation>, Serializable { // for Spark jobs
+public final class Annotation implements Comparable<Annotation>, Serializable { // for Spark jobs
   private static final long serialVersionUID = 0L;
 
   public static Annotation create(long timestamp, String value) {
-    return new AutoValue_Annotation(timestamp, value);
+    if (value == null) throw new NullPointerException("value == null");
+    return new Annotation(timestamp, value);
   }
 
   /**
@@ -36,12 +37,17 @@ public abstract class Annotation implements Comparable<Annotation>, Serializable
    * <p>This value should be set directly by instrumentation, using the most precise value possible.
    * For example, {@code gettimeofday} or multiplying {@link System#currentTimeMillis} by 1000.
    */
-  public abstract long timestamp();
+  public long timestamp() {
+    return timestamp;
+  }
 
   /**
    * Usually a short tag indicating an event, like {@code cache.miss} or {@code error}
    */
-  public abstract String value();
+  public String value() {
+    return value;
+  }
+
 
   /** Compares by {@link #timestamp}, then {@link #value}. */
   @Override public int compareTo(Annotation that) {
@@ -51,6 +57,61 @@ public abstract class Annotation implements Comparable<Annotation>, Serializable
     return value().compareTo(that.value());
   }
 
-  Annotation() {
+  // clutter below mainly due to difficulty working with Kryo which cannot handle AutoValue subclass
+  // See https://github.com/openzipkin/zipkin/issues/1879
+  final long timestamp;
+  final String value;
+
+  Annotation(long timestamp, String value) {
+    this.timestamp = timestamp;
+    this.value = value;
+  }
+
+  @Override public String toString() {
+    return "Annotation{"
+      + "timestamp=" + timestamp + ", "
+      + "value=" + value
+      + "}";
+  }
+
+  @Override public boolean equals(Object o) {
+    if (o == this) return true;
+    if (!(o instanceof Annotation)) return false;
+    Annotation that = (Annotation) o;
+    return (timestamp == that.timestamp()) && (value.equals(that.value()));
+  }
+
+  @Override public int hashCode() {
+    int h = 1;
+    h *= 1000003;
+    h ^= (int) ((timestamp >>> 32) ^ timestamp);
+    h *= 1000003;
+    h ^= value.hashCode();
+    return h;
+  }
+
+  // As this is an immutable object (no default constructor), defer to a serialization proxy.
+  final Object writeReplace() throws ObjectStreamException {
+    return new SerializedForm(this);
+  }
+
+  private static final class SerializedForm implements Serializable {
+    static final long serialVersionUID = 0L;
+
+    final long timestamp;
+    final String value;
+
+    SerializedForm(Annotation annotation) {
+      timestamp = annotation.timestamp;
+      value = annotation.value;
+    }
+
+    Object readResolve() throws ObjectStreamException {
+      try {
+        return Annotation.create(timestamp, value);
+      } catch (IllegalArgumentException e) {
+        throw new StreamCorruptedException(e.getMessage());
+      }
+    }
   }
 }

--- a/zipkin2/src/main/java/zipkin2/DependencyLink.java
+++ b/zipkin2/src/main/java/zipkin2/DependencyLink.java
@@ -1,5 +1,5 @@
 /**
- * Copyright 2015-2017 The OpenZipkin Authors
+ * Copyright 2015-2018 The OpenZipkin Authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
  * in compliance with the License. You may obtain a copy of the License at
@@ -13,60 +13,90 @@
  */
 package zipkin2;
 
-import com.google.auto.value.AutoValue;
+import java.io.ObjectStreamException;
 import java.io.Serializable;
+import java.io.StreamCorruptedException;
 import java.nio.charset.Charset;
 import java.util.Locale;
+import zipkin2.codec.DependencyLinkBytesDecoder;
 import zipkin2.codec.DependencyLinkBytesEncoder;
 
-@AutoValue
 //@Immutable
-public abstract class DependencyLink implements Serializable { // for Spark jobs
+public final class DependencyLink implements Serializable { // for Spark and Flink jobs
   static final Charset UTF_8 = Charset.forName("UTF-8");
 
   private static final long serialVersionUID = 0L;
 
   public static Builder newBuilder() {
-    return new AutoValue_DependencyLink.Builder().errorCount(0);
+    return new Builder();
   }
 
   /** parent service name (caller) */
-  public abstract String parent();
+  public String parent() {
+    return parent;
+  }
 
   /** child service name (callee) */
-  public abstract String child();
+  public String child() {
+    return child;
+  }
 
   /** total traced calls made from {@link #parent} to {@link #child} */
-  public abstract long callCount();
+  public long callCount() {
+    return callCount;
+  }
 
   /** How many {@link #callCount calls} are known to be errors */
-  public abstract long errorCount();
+  public long errorCount() {
+    return errorCount;
+  }
 
-  public abstract Builder toBuilder();
+  public Builder toBuilder() {
+    return new Builder(this);
+  }
 
-  @AutoValue.Builder
-  public static abstract class Builder {
-
-    public abstract Builder parent(String parent);
-
-    public abstract Builder child(String child);
-
-    public abstract Builder callCount(long callCount);
-
-    public abstract Builder errorCount(long errorCount);
-
-    abstract String parent();
-
-    abstract String child();
-
-    abstract DependencyLink autoBuild();
-
-    public final DependencyLink build() {
-      return parent(parent().toLowerCase(Locale.ROOT))
-        .child(child().toLowerCase(Locale.ROOT)).autoBuild();
-    }
+  public static final class Builder {
+    String parent, child;
+    long callCount, errorCount;
 
     Builder() {
+    }
+
+    Builder(DependencyLink source) {
+      this.parent = source.parent;
+      this.child = source.child;
+      this.callCount = source.callCount;
+      this.errorCount = source.errorCount;
+    }
+
+    public Builder parent(String parent) {
+      if (parent == null) throw new NullPointerException("parent == null");
+      this.parent = parent.toLowerCase(Locale.ROOT);
+      return this;
+    }
+
+    public Builder child(String child) {
+      if (child == null) throw new NullPointerException("child == null");
+      this.child = child.toLowerCase(Locale.ROOT);
+      return this;
+    }
+
+    public Builder callCount(long callCount) {
+      this.callCount = callCount;
+      return this;
+    }
+
+    public Builder errorCount(long errorCount) {
+      this.errorCount = errorCount;
+      return this;
+    }
+
+    public DependencyLink build() {
+      String missing = "";
+      if (parent == null) missing += " parent";
+      if (child == null) missing += " child";
+      if (!"".equals(missing)) throw new IllegalStateException("Missing :" + missing);
+      return new DependencyLink(this);
     }
   }
 
@@ -74,6 +104,61 @@ public abstract class DependencyLink implements Serializable { // for Spark jobs
     return new String(DependencyLinkBytesEncoder.JSON_V1.encode(this), UTF_8);
   }
 
-  DependencyLink() {
+  // clutter below mainly due to difficulty working with Kryo which cannot handle AutoValue subclass
+  // See https://github.com/openzipkin/zipkin/issues/1879
+  final String parent, child;
+  final long callCount, errorCount;
+
+  DependencyLink(Builder builder) {
+    parent = builder.parent;
+    child = builder.child;
+    callCount = builder.callCount;
+    errorCount = builder.errorCount;
+  }
+
+  @Override public boolean equals(Object o) {
+    if (o == this) return true;
+    if (!(o instanceof DependencyLink)) return false;
+    DependencyLink that = (DependencyLink) o;
+      return (parent.equals(that.parent))
+        && (child.equals(that.child))
+        && (callCount == that.callCount)
+        && (errorCount == that.errorCount);
+  }
+
+  @Override public int hashCode() {
+    int h = 1;
+    h *= 1000003;
+    h ^= parent.hashCode();
+    h *= 1000003;
+    h ^= child.hashCode();
+    h *= 1000003;
+    h ^= (int) ((callCount >>> 32) ^ callCount);
+    h *= 1000003;
+    h ^= (int) ((errorCount >>> 32) ^ errorCount);
+    return h;
+  }
+
+  // This is an immutable object, and our encoder is faster than java's: use a serialization proxy.
+  final Object writeReplace() throws ObjectStreamException {
+    return new SerializedForm(DependencyLinkBytesEncoder.JSON_V1.encode(this));
+  }
+
+  private static final class SerializedForm implements Serializable {
+    private static final long serialVersionUID = 0L;
+
+    final byte[] bytes;
+
+    SerializedForm(byte[] bytes) {
+      this.bytes = bytes;
+    }
+
+    Object readResolve() throws ObjectStreamException {
+      try {
+        return DependencyLinkBytesDecoder.JSON_V1.decodeOne(bytes);
+      } catch (IllegalArgumentException e) {
+        throw new StreamCorruptedException(e.getMessage());
+      }
+    }
   }
 }

--- a/zipkin2/src/main/java/zipkin2/Endpoint.java
+++ b/zipkin2/src/main/java/zipkin2/Endpoint.java
@@ -1,5 +1,5 @@
 /**
- * Copyright 2015-2017 The OpenZipkin Authors
+ * Copyright 2015-2018 The OpenZipkin Authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
  * in compliance with the License. You may obtain a copy of the License at
@@ -13,8 +13,9 @@
  */
 package zipkin2;
 
-import com.google.auto.value.AutoValue;
+import java.io.ObjectStreamException;
 import java.io.Serializable;
+import java.io.StreamCorruptedException;
 import java.net.Inet4Address;
 import java.net.Inet6Address;
 import java.net.InetAddress;
@@ -23,9 +24,8 @@ import java.util.Locale;
 import zipkin2.internal.Nullable;
 
 /** The network context of a node in the service graph. */
-@AutoValue
 //@Immutable
-public abstract class Endpoint implements Serializable { // for Spark jobs
+public final class Endpoint implements Serializable { // for Spark and Flink jobs
   private static final long serialVersionUID = 0L;
 
   /**
@@ -35,13 +35,17 @@ public abstract class Endpoint implements Serializable { // for Spark jobs
    * <p>This is a primary label for trace lookup and aggregation, so it should be intuitive and
    * consistent. Many use a name from service discovery.
    */
-  @Nullable public abstract String serviceName();
+  @Nullable public String serviceName() {
+    return serviceName;
+  }
 
   /**
    * The text representation of the primary IPv4 address associated with this a connection. Ex.
    * 192.168.99.100 Absent if unknown.
    */
-  @Nullable public abstract String ipv4();
+  @Nullable public String ipv4() {
+    return ipv4;
+  }
 
   /**
    * The text representation of the primary IPv6 address associated with this a connection. Ex.
@@ -49,25 +53,44 @@ public abstract class Endpoint implements Serializable { // for Spark jobs
    *
    * <p>Prefer using the {@link #ipv4()} field for mapped addresses.
    */
-  @Nullable public abstract String ipv6();
+  @Nullable public String ipv6() {
+    return ipv6;
+  }
 
   /**
    * Port of the IP's socket or null, if not known.
    *
    * @see java.net.InetSocketAddress#getPort()
    */
-  @Nullable public abstract Integer port();
-
-  public abstract Builder toBuilder();
-
-  public static Builder newBuilder() {
-    return new AutoValue_Endpoint.Builder();
+  @Nullable public Integer port() {
+    return port;
   }
 
-  @AutoValue.Builder
-  public static abstract class Builder {
+  public Builder toBuilder() {
+    return new Builder(this);
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  public static final class Builder {
+    String serviceName, ipv4, ipv6;
+    Integer port;
+
+    Builder(Endpoint source) {
+      serviceName = source.serviceName;
+      ipv4 = source.ipv4;
+      ipv6 = source.ipv6;
+      port = source.port;
+    }
+
     /** @see Endpoint#serviceName */
-    public abstract Builder serviceName(@Nullable String serviceName);
+    public Builder serviceName(@Nullable String serviceName) {
+      this.serviceName = serviceName == null || serviceName.isEmpty()
+        ? null : serviceName.toLowerCase(Locale.ROOT);
+      return this;
+    }
 
     /** Chaining variant of {@link #parseIp(InetAddress)} */
     public Builder ip(@Nullable InetAddress addr) {
@@ -76,7 +99,7 @@ public abstract class Endpoint implements Serializable { // for Spark jobs
     }
 
     /**
-     * Returns true if {@link #ipv4(String)} or {@link #ipv6(String)} could be parsed from the
+     * Returns true if {@link Endpoint#ipv4()} or {@link Endpoint#ipv6()} could be parsed from the
      * input.
      *
      * <p>Returns boolean not this for conditional parsing. For example:
@@ -91,14 +114,14 @@ public abstract class Endpoint implements Serializable { // for Spark jobs
     public final boolean parseIp(@Nullable InetAddress addr) {
       if (addr == null) return false;
       if (addr instanceof Inet4Address) {
-        ipv4(addr.getHostAddress());
+        ipv4 = addr.getHostAddress();
       } else if (addr instanceof Inet6Address) {
         byte[] addressBytes = addr.getAddress();
         String ipv4 = parseEmbeddedIPv4(addressBytes);
         if (ipv4 != null) {
-          ipv4(ipv4);
+          this.ipv4 = ipv4;
         } else {
-          ipv6(writeIpV6(addressBytes));
+          ipv6 = writeIpV6(addressBytes);
         }
       } else {
         return false;
@@ -113,7 +136,7 @@ public abstract class Endpoint implements Serializable { // for Spark jobs
     }
 
     /**
-     * Returns true if {@link #ipv4(String)} or {@link #ipv6(String)} could be parsed from the
+     * Returns true if {@link Endpoint#ipv4()} or {@link Endpoint#ipv6()} could be parsed from the
      * input.
      *
      * <p>Returns boolean not this for conditional parsing. For example:
@@ -129,27 +152,18 @@ public abstract class Endpoint implements Serializable { // for Spark jobs
       if (ipString == null || ipString.isEmpty()) return false;
       IpFamily format = detectFamily(ipString);
       if (format == IpFamily.IPv4) {
-        ipv4(ipString);
+        ipv4 = ipString;
       } else if (format == IpFamily.IPv4Embedded) {
-        ipv4(ipString.substring(ipString.lastIndexOf(':') + 1));
+        ipv4 = ipString.substring(ipString.lastIndexOf(':') + 1);
       } else if (format == IpFamily.IPv6) {
         byte[] addressBytes = textToNumericFormatV6(ipString);
         if (addressBytes == null) return false;
-        ipv6(writeIpV6(addressBytes)); // ensures consistent format
+        ipv6 = writeIpV6(addressBytes); // ensures consistent format
       } else {
         return false;
       }
       return true;
     }
-
-    // hidden to ensure input are parsed as otherwise equals comparisons like needed for clock skew
-    // won't work.
-
-    /** @see Endpoint#ipv4 */
-    abstract Builder ipv4(@Nullable String ipv4);
-
-    /** @see Endpoint#ipv4 */
-    abstract Builder ipv6(@Nullable String ipv6);
 
     /**
      * Use this to set the port to an externally defined value.
@@ -157,29 +171,17 @@ public abstract class Endpoint implements Serializable { // for Spark jobs
      * @param port port associated with the endpoint. zero coerces to null (unknown)
      * @see Endpoint#port()
      */
-    public abstract Builder port(@Nullable Integer port);
-
-    abstract @Nullable String serviceName();
-
-    abstract @Nullable Integer port();
-
-    abstract Endpoint autoBuild();
-
-    public final Endpoint build() {
-      String serviceName = serviceName();
-      if (serviceName != null) {
-        if (serviceName.isEmpty()) {
-          serviceName(null);
-        } else {
-          serviceName(serviceName.toLowerCase(Locale.ROOT));
-        }
-      }
-      Integer port = port();
+    public Builder port(@Nullable Integer port) {
       if (port != null) {
         if (port > 0xffff) throw new IllegalArgumentException("invalid port " + port);
-        if (port <= 0) port(null);
+        if (port <= 0) port = null;
       }
-      return autoBuild();
+      this.port = port;
+      return this;
+    }
+
+    public Endpoint build() {
+      return new Endpoint(this);
     }
 
     Builder() {
@@ -434,4 +436,83 @@ public abstract class Endpoint implements Serializable { // for Spark jobs
     return c >= '0' && c <= '9';
   }
   // End code from io.netty.util.NetUtil 4.1
+
+  // clutter below mainly due to difficulty working with Kryo which cannot handle AutoValue subclass
+  // See https://github.com/openzipkin/zipkin/issues/1879
+  final String serviceName, ipv4, ipv6;
+  final Integer port;
+
+  Endpoint(Builder builder) {
+    serviceName = builder.serviceName;
+    ipv4 = builder.ipv4;
+    ipv6 = builder.ipv6;
+    port = builder.port;
+  }
+
+  Endpoint(SerializedForm serializedForm) {
+    serviceName = serializedForm.serviceName;
+    ipv4 = serializedForm.ipv4;
+    ipv6 = serializedForm.ipv6;
+    port = serializedForm.port;
+  }
+
+  @Override public String toString() {
+    return "Endpoint{"
+      + "serviceName=" + serviceName + ", "
+      + "ipv4=" + ipv4 + ", "
+      + "ipv6=" + ipv6 + ", "
+      + "port=" + port
+      + "}";
+  }
+
+  @Override public boolean equals(Object o) {
+    if (o == this) return true;
+    if (!(o instanceof Endpoint)) return false;
+    Endpoint that = (Endpoint) o;
+    return ((serviceName == null)
+      ? (that.serviceName == null) : serviceName.equals(that.serviceName))
+      && ((ipv4 == null) ? (that.ipv4 == null) : ipv4.equals(that.ipv4))
+      && ((ipv6 == null) ? (that.ipv6 == null) : ipv6.equals(that.ipv6))
+      && ((port == null) ? (that.port == null) : port.equals(that.port));
+  }
+
+  @Override public int hashCode() {
+    int h = 1;
+    h *= 1000003;
+    h ^= (serviceName == null) ? 0 : serviceName.hashCode();
+    h *= 1000003;
+    h ^= (ipv4 == null) ? 0 : ipv4.hashCode();
+    h *= 1000003;
+    h ^= (ipv6 == null) ? 0 : ipv6.hashCode();
+    h *= 1000003;
+    h ^= (port == null) ? 0 : port.hashCode();
+    return h;
+  }
+
+  // As this is an immutable object (no default constructor), defer to a serialization proxy.
+  final Object writeReplace() throws ObjectStreamException {
+    return new SerializedForm(this);
+  }
+
+  private static final class SerializedForm implements Serializable {
+    static final long serialVersionUID = 0L;
+
+    final String serviceName, ipv4, ipv6;
+    final Integer port;
+
+    SerializedForm(Endpoint endpoint) {
+      serviceName = endpoint.serviceName;
+      ipv4 = endpoint.ipv4;
+      ipv6 = endpoint.ipv6;
+      port = endpoint.port;
+    }
+
+    Object readResolve() throws ObjectStreamException {
+      try {
+        return new Endpoint(this);
+      } catch (IllegalArgumentException e) {
+        throw new StreamCorruptedException(e.getMessage());
+      }
+    }
+  }
 }

--- a/zipkin2/src/main/java/zipkin2/Span.java
+++ b/zipkin2/src/main/java/zipkin2/Span.java
@@ -1,5 +1,5 @@
 /**
- * Copyright 2015-2017 The OpenZipkin Authors
+ * Copyright 2015-2018 The OpenZipkin Authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
  * in compliance with the License. You may obtain a copy of the License at
@@ -13,8 +13,9 @@
  */
 package zipkin2;
 
-import com.google.auto.value.AutoValue;
+import java.io.ObjectStreamException;
 import java.io.Serializable;
+import java.io.StreamCorruptedException;
 import java.nio.charset.Charset;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -24,8 +25,9 @@ import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 import java.util.TreeMap;
-import zipkin2.internal.Nullable;
+import zipkin2.codec.SpanBytesDecoder;
 import zipkin2.codec.SpanBytesEncoder;
+import zipkin2.internal.Nullable;
 
 /**
  * A trace is a series of spans (often RPC calls) which form a latency tree.
@@ -47,9 +49,8 @@ import zipkin2.codec.SpanBytesEncoder;
  * not need to repeat endpoints on each data like {@code zipkin.Span} does. This results in simpler
  * and smaller data.
  */
-@AutoValue
 //@Immutable
-public abstract class Span implements Serializable { // for Spark jobs
+public final class Span implements Serializable { // for Spark and Flink jobs
   static final Charset UTF_8 = Charset.forName("UTF-8");
 
   private static final long serialVersionUID = 0L;
@@ -63,14 +64,18 @@ public abstract class Span implements Serializable { // for Spark jobs
    * <p>Some systems downgrade trace identifiers to 64bit by dropping the left-most 16 characters.
    * For example, {@code 4e441824ec2b6a44ffdc9bb9a6453df3} becomes {@code ffdc9bb9a6453df3}.
    */
-  public abstract String traceId();
+  public String traceId() {
+    return traceId;
+  }
 
   /**
    * The parent's {@link #id} or null if this the root span in a trace.
    *
    * <p>This is the same encoding as {@link #id}. For example {@code ffdc9bb9a6453df3}
    */
-  @Nullable public abstract String parentId();
+  @Nullable public String parentId() {
+    return parentId;
+  }
 
   /**
    * Unique 64bit identifier for this operation within the trace.
@@ -79,7 +84,9 @@ public abstract class Span implements Serializable { // for Spark jobs
    *
    * <p>A span is uniquely identified in storage by ({@linkplain #traceId}, {@linkplain #id()}).
    */
-  public abstract String id();
+  public String id() {
+    return id;
+  }
 
   /** Indicates the primary span type. */
   public enum Kind {
@@ -106,14 +113,18 @@ public abstract class Span implements Serializable { // for Spark jobs
   }
 
   /** When present, used to interpret {@link #remoteEndpoint} */
-  @Nullable public abstract Kind kind();
+  @Nullable public Kind kind() {
+    return kind;
+  }
 
   /**
    * Span name in lowercase, rpc method for example.
    *
    * <p>Conventionally, when the span name isn't known, name = "unknown".
    */
-  @Nullable public abstract String name();
+  @Nullable public String name() {
+    return name;
+  }
 
   /**
    * Epoch microseconds of the start of this span, possibly absent if this an incomplete span.
@@ -131,7 +142,9 @@ public abstract class Span implements Serializable { // for Spark jobs
    *
    * @see #duration()
    */
-  @Nullable public abstract Long timestamp();
+  @Nullable public Long timestamp() {
+    return timestamp;
+  }
 
   /**
    * Measurement in microseconds of the critical path, if known. Durations of less than one
@@ -147,7 +160,9 @@ public abstract class Span implements Serializable { // for Spark jobs
    *
    * <p>This field is i64 vs i32 to support spans longer than 35 minutes.
    */
-  @Nullable public abstract Long duration();
+  @Nullable public Long duration() {
+    return duration;
+  }
 
   /**
    * The host that recorded this span, primarily for query by service name.
@@ -156,7 +171,9 @@ public abstract class Span implements Serializable { // for Spark jobs
    * name as it is used in search. This is nullable for legacy reasons.
    */
   // Nullable for data conversion especially late arriving data which might not have an annotation
-  @Nullable public abstract Endpoint localEndpoint();
+  @Nullable public Endpoint localEndpoint() {
+    return localEndpoint;
+  }
 
   /**
    * When an RPC (or messaging) span, indicates the other side of the connection.
@@ -165,24 +182,32 @@ public abstract class Span implements Serializable { // for Spark jobs
    * is not tracing. For example,For example, you can record the IP from the {@code X-Forwarded-For}
    * header or or the service name and socket of a remote peer.
    */
-  @Nullable public abstract Endpoint remoteEndpoint();
+  @Nullable public Endpoint remoteEndpoint() {
+    return remoteEndpoint;
+  }
 
   /**
    * Events that explain latency with a timestamp. Unlike log statements, annotations are often
    * short or contain codes: for example "brave.flush". Annotations are sorted ascending by
    * timestamp.
    */
-  public abstract List<Annotation> annotations();
+  public List<Annotation> annotations() {
+    return annotations;
+  }
 
   /**
    * Tags a span with context, usually to support query or aggregation.
    *
    * <p>For example, a tag key could be {@code "http.path"}.
    */
-  public abstract Map<String, String> tags();
+  public Map<String, String> tags() {
+    return tags;
+  }
 
   /** True is a request to store this span even if it overrides sampling policy. */
-  @Nullable public abstract Boolean debug();
+  @Nullable public Boolean debug(){
+    return debug;
+  }
 
   /**
    * True if we are contributing to a span started by another tracer (ex on a different host).
@@ -192,7 +217,9 @@ public abstract class Span implements Serializable { // for Spark jobs
    * the server side. However, the server shouldn't set span.timestamp or duration since it didn't
    * start the span.
    */
-  @Nullable public abstract Boolean shared();
+  @Nullable public Boolean shared() {
+    return shared;
+  }
 
   @Nullable public String localServiceName() {
     Endpoint localEndpoint = localEndpoint();
@@ -213,19 +240,14 @@ public abstract class Span implements Serializable { // for Spark jobs
   }
 
   public static final class Builder {
-    String traceId;
-    String parentId;
-    String id;
+    String traceId, parentId, id;
     Kind kind;
     String name;
-    Long timestamp;
-    Long duration;
-    Endpoint localEndpoint;
-    Endpoint remoteEndpoint;
+    Long timestamp, duration;
+    Endpoint localEndpoint, remoteEndpoint;
     ArrayList<Annotation> annotations;
     TreeMap<String, String> tags;
-    Boolean debug;
-    Boolean shared;
+    Boolean debug, shared;
 
     public Builder clear() {
       traceId = null;
@@ -267,25 +289,25 @@ public abstract class Span implements Serializable { // for Spark jobs
     }
 
     Builder(Span source) {
-      traceId = source.traceId();
-      parentId = source.parentId();
-      id = source.id();
-      kind = source.kind();
-      name = source.name();
-      timestamp = source.timestamp();
-      duration = source.duration();
-      localEndpoint = source.localEndpoint();
-      remoteEndpoint = source.remoteEndpoint();
-      if (!source.annotations().isEmpty()) {
-        annotations = new ArrayList<>(source.annotations().size());
-        annotations.addAll(source.annotations());
+      traceId = source.traceId;
+      parentId = source.parentId;
+      id = source.id;
+      kind = source.kind;
+      name = source.name;
+      timestamp = source.timestamp;
+      duration = source.duration;
+      localEndpoint = source.localEndpoint;
+      remoteEndpoint = source.remoteEndpoint;
+      if (!source.annotations.isEmpty()) {
+        annotations = new ArrayList<>(source.annotations.size());
+        annotations.addAll(source.annotations);
       }
-      if (!source.tags().isEmpty()) {
+      if (!source.tags.isEmpty()) {
         tags = new TreeMap<>();
-        tags.putAll(source.tags());
+        tags.putAll(source.tags);
       }
-      debug = source.debug();
-      shared = source.shared();
+      debug = source.debug;
+      shared = source.shared;
     }
 
     @Nullable public Kind kind() {
@@ -401,21 +423,11 @@ public abstract class Span implements Serializable { // for Spark jobs
     }
 
     public Span build() {
-      return new AutoValue_Span(
-        traceId,
-        parentId,
-        id,
-        kind,
-        name,
-        timestamp,
-        duration,
-        localEndpoint,
-        remoteEndpoint,
-        sortedList(annotations),
-        tags == null ? Collections.emptyMap() : new LinkedHashMap<>(tags),
-        debug,
-        shared
-      );
+      String missing = "";
+      if (traceId == null) missing += " traceId";
+      if (id == null) missing += " id";
+      if (!"".equals(missing)) throw new IllegalStateException("Missing :" + missing);
+      return new Span(this);
     }
 
     Builder() {
@@ -470,5 +482,107 @@ public abstract class Span implements Serializable { // for Spark jobs
     Arrays.sort(array);
     List result = Arrays.asList(array);
     return Collections.unmodifiableList(result);
+  }
+
+  // clutter below mainly due to difficulty working with Kryo which cannot handle AutoValue subclass
+  // See https://github.com/openzipkin/zipkin/issues/1879
+  final String traceId, parentId, id;
+  final Kind kind;
+  final String name;
+  final Long timestamp, duration;
+  final Endpoint localEndpoint, remoteEndpoint;
+  final List<Annotation> annotations;
+  final Map<String, String> tags;
+  final Boolean debug, shared;
+
+  Span(Builder builder) {
+    traceId = builder.traceId;
+    parentId = builder.parentId;
+    id = builder.id;
+    kind = builder.kind;
+    name = builder.name;
+    timestamp = builder.timestamp;
+    duration = builder.duration;
+    localEndpoint = builder.localEndpoint;
+    remoteEndpoint = builder.remoteEndpoint;
+    annotations = sortedList(builder.annotations);
+    tags = builder.tags == null ? Collections.emptyMap() : new LinkedHashMap<>(builder.tags);
+    debug = builder.debug;
+    shared = builder.shared;
+  }
+
+  @Override public boolean equals(Object o) {
+    if (o == this) return true;
+    if (!(o instanceof Span)) return false;
+    Span that = (Span) o;
+    return (traceId.equals(that.traceId))
+      && ((parentId == null) ? (that.parentId == null) : parentId.equals(that.parentId))
+      && (id.equals(that.id))
+      && ((kind == null) ? (that.kind == null) : kind.equals(that.kind))
+      && ((name == null) ? (that.name == null) : name.equals(that.name))
+      && ((timestamp == null) ? (that.timestamp == null) : timestamp.equals(that.timestamp))
+      && ((duration == null) ? (that.duration == null) : duration.equals(that.duration))
+      && ((localEndpoint == null)
+      ? (that.localEndpoint == null) : localEndpoint.equals(that.localEndpoint))
+      && ((remoteEndpoint == null)
+      ? (that.remoteEndpoint == null) : remoteEndpoint.equals(that.remoteEndpoint))
+      && (annotations.equals(that.annotations))
+      && (tags.equals(that.tags))
+      && ((debug == null) ? (that.debug == null) : debug.equals(that.debug))
+      && ((shared == null) ? (that.shared == null) : shared.equals(that.shared));
+  }
+
+  @Override public int hashCode() {
+    int h = 1;
+    h *= 1000003;
+    h ^= traceId.hashCode();
+    h *= 1000003;
+    h ^= (parentId == null) ? 0 : parentId.hashCode();
+    h *= 1000003;
+    h ^= id.hashCode();
+    h *= 1000003;
+    h ^= (kind == null) ? 0 : kind.hashCode();
+    h *= 1000003;
+    h ^= (name == null) ? 0 : name.hashCode();
+    h *= 1000003;
+    h ^= (timestamp == null) ? 0 : timestamp.hashCode();
+    h *= 1000003;
+    h ^= (duration == null) ? 0 : duration.hashCode();
+    h *= 1000003;
+    h ^= (localEndpoint == null) ? 0 : localEndpoint.hashCode();
+    h *= 1000003;
+    h ^= (remoteEndpoint == null) ? 0 : remoteEndpoint.hashCode();
+    h *= 1000003;
+    h ^= annotations.hashCode();
+    h *= 1000003;
+    h ^= tags.hashCode();
+    h *= 1000003;
+    h ^= (debug == null) ? 0 : debug.hashCode();
+    h *= 1000003;
+    h ^= (shared == null) ? 0 : shared.hashCode();
+    return h;
+  }
+
+  // This is an immutable object, and our encoder is faster than java's: use a serialization proxy.
+  final Object writeReplace() throws ObjectStreamException {
+    return new SerializedForm(SpanBytesEncoder.JSON_V2.encode(this));
+  }
+
+  private static final class SerializedForm implements Serializable {
+    private static final long serialVersionUID = 0L;
+
+    final byte[] bytes;
+
+    SerializedForm(byte[] bytes) {
+      this.bytes = bytes;
+    }
+
+    Object readResolve() throws ObjectStreamException {
+      try {
+        return SpanBytesDecoder.JSON_V2.decodeOne(bytes);
+      } catch (IllegalArgumentException e) {
+        throw new StreamCorruptedException(e.getMessage());
+      }
+    }
   }
 }

--- a/zipkin2/src/test/java/zipkin2/EndpointTest.java
+++ b/zipkin2/src/test/java/zipkin2/EndpointTest.java
@@ -1,5 +1,5 @@
 /**
- * Copyright 2015-2017 The OpenZipkin Authors
+ * Copyright 2015-2018 The OpenZipkin Authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
  * in compliance with the License. You may obtain a copy of the License at
@@ -109,7 +109,7 @@ public class EndpointTest {
       .isNull();
   }
 
-  @Test public void ip_ipv6_compatIpv4() throws UnknownHostException {
+  @Test public void ip_ipv6_compatIpv4() {
     String ipv6 = "::0000:1.2.3.4";
     Endpoint endpoint = Endpoint.newBuilder().ip(ipv6).build();
 
@@ -195,7 +195,7 @@ public class EndpointTest {
   }
 
   @Test public void lowercasesServiceName() {
-    assertThat(Endpoint.newBuilder().serviceName("fFf").ipv4("127.0.0.1").build().serviceName())
+    assertThat(Endpoint.newBuilder().serviceName("fFf").ip("127.0.0.1").build().serviceName())
       .isEqualTo("fff");
   }
 }

--- a/zipkin2/src/test/java/zipkin2/codec/KryoTest.java
+++ b/zipkin2/src/test/java/zipkin2/codec/KryoTest.java
@@ -1,0 +1,114 @@
+/**
+ * Copyright 2015-2018 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.codec;
+
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.Serializer;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import com.esotericsoftware.kryo.serializers.JavaSerializer;
+import org.junit.Test;
+import zipkin2.Annotation;
+import zipkin2.DependencyLink;
+import zipkin2.Endpoint;
+import zipkin2.Span;
+import zipkin2.TestObjects;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class KryoTest {
+
+  @Test public void kryoJavaSerialization_annotation() {
+    Kryo kryo = new Kryo();
+    kryo.register(Annotation.class, new JavaSerializer());
+
+    Output output = new Output(4096);
+    kryo.writeObject(output, Annotation.create(1L, "foo"));
+    output.flush();
+    byte[] serialized = output.getBuffer();
+
+    assertThat(kryo.readObject(new Input(serialized), Annotation.class))
+      .isEqualTo(Annotation.create(1L, "foo"));
+  }
+
+  @Test public void kryoJavaSerialization_endpoint() {
+    Kryo kryo = new Kryo();
+    kryo.register(Endpoint.class, new JavaSerializer());
+
+    Output output = new Output(4096);
+    kryo.writeObject(output, TestObjects.BACKEND);
+    output.flush();
+    byte[] serialized = output.getBuffer();
+
+    assertThat(kryo.readObject(new Input(serialized), Endpoint.class))
+      .isEqualTo(TestObjects.BACKEND);
+  }
+
+  @Test public void kryoJavaSerialization_span() {
+    Kryo kryo = new Kryo();
+    kryo.register(Span.class, new JavaSerializer());
+
+    Output output = new Output(4096);
+    kryo.writeObject(output, TestObjects.CLIENT_SPAN);
+    output.flush();
+    byte[] serialized = output.getBuffer();
+
+    assertThat(kryo.readObject(new Input(serialized), Span.class))
+      .isEqualTo(TestObjects.CLIENT_SPAN);
+  }
+
+  @Test public void kryoJavaSerialization_dependencyLink() {
+    Kryo kryo = new Kryo();
+    kryo.register(DependencyLink.class, new JavaSerializer());
+
+    DependencyLink link = DependencyLink.newBuilder().parent("client").child("server").callCount(2L)
+      .errorCount(23L).build();
+
+    Output output = new Output(4096);
+    kryo.writeObject(output, link);
+    output.flush();
+    byte[] serialized = output.getBuffer();
+
+    assertThat(kryo.readObject(new Input(serialized), DependencyLink.class))
+      .isEqualTo(link);
+  }
+
+  /** Example test for how to use Kryo and reuse our encoders */
+  public class JsonV2SpanSerializer extends Serializer<Span> {
+    @Override public void write(Kryo kryo, Output output, Span span) {
+      byte[] json = SpanBytesEncoder.JSON_V2.encode(span);
+      output.writeInt(json.length);
+      output.write(json);
+    }
+
+    @Override public Span read(Kryo kryo, Input input, Class<Span> type) {
+      int length = input.readInt();
+      byte[] json = input.readBytes(length);
+      return SpanBytesDecoder.JSON_V2.decodeOne(json);
+    }
+  }
+
+  @Test public void kryoJson2() {
+    Kryo kryo = new Kryo();
+    kryo.register(Span.class, new JsonV2SpanSerializer());
+
+    Output output = new Output(4096);
+    kryo.writeObject(output, TestObjects.CLIENT_SPAN);
+    output.flush();
+    byte[] serialized = output.getBuffer();
+
+    assertThat(kryo.readObject(new Input(serialized), Span.class))
+      .isEqualTo(TestObjects.CLIENT_SPAN);
+  }
+}

